Grate 3d API, textures, refactoring and some fixes

digetx opened the pull request:
- Added a basic 3d context API to grate, so whacking new tests and features should be less tedious.
- We have now a unified way of describing framebuffer/render targets/textures - pixel buffer.
- Cube-textured grate test.
- Couple new minor features like rgb565 framebuffer support and detiling using gr2d.
- Grate single buffered output is fixed, libwrap fix, inverted nvhost display fixed, DRM VSYNC fixes.

digetx committed:
libcgc: compile all sources unconditionally

Provide BLOB stubs instead of the libcgc's.

digetx committed:
import tgr_3d header generated by enytools

The plan is to use as the source of registers and bitfields literals.

digetx committed:
grate: add grate_3d_*() helpers

It is more readable now. Start to use envytools-generated header.

digetx committed:
partial type conversion cleanup

I have been bitten several times by it and decided to cleanup, so we
could have -Wconversion in CFLAGS . Unfortunately, it's just too many
places to fix, I gave up.

digetx committed:
tests/grate: provide linker asm for the tests

That's more consistent and now we can drop the hardcoded linker code from
the grate_draw_elements().

digetx committed:
grate: derive asm related functions from shader-cgc.c to grate-asm.c

The shader-cgc got bloated too much, it's easier to navigate the separate files.

digetx committed:
libwrap: fix argument-less ioctl call

The _IOC_SIZE() macro doesn't represent the presence of an argument. There is
always an argument, even when it is absent, i.e. NULL.

digetx committed:
grate: factor out 3d related functions to grate-3d.c

Logical split, no functional change.

digetx committed:
grate: factor out 2d related functions to grate-2d.c

Logical split, no functional change.

digetx committed:
grate: add 3d context API

Get rid of hardcoded 3d context params. It should be much easier now to
implement textures and other goodies. Double buffered display output is gone
from grate tests for now, while single got fixed.

Sync envytools-generated tgr_3d.xml.h

digetx committed:
sync envytools-generated tgr_3d.xml.h

digetx committed:
grate: initial texture support

Basic texture support, exposed via 3d context API.

digetx committed:
tests/grate: add cube-textured test

New build dependency - libdevil.

digetx committed:
host1x: introduce pixel buffer

The intention is to use it everywhere as a universal pixel storage.

digetx committed:
grate: support RGB565 framebuffer format

libhost1x could handle it now.

digetx committed:
libhost1x: invert vertically nvhost display output

The 3d render is not inverted now, DRM does the same.

digetx committed:
tests/grate: cube-textured: invert texture on load

So it is not inverted and matches the opengl convention.

digetx committed:
libhost1x: provide CRTC pipe number on overlay VBLANK

VSYNC works with a non-primary CRTC.

digetx committed:
libhost1x: try to change display mode on drmModePageFlip() error

Page flip IOCTL forbids changing framebuffer format, so format needs to be
changed first.

digetx committed:
libhost1x: make tiled display format optional

digetx committed:
grate: drop grate_format in favor of pixel_format

The non-tiled framebuffer format could be set directly now, grate_format
is unneeded intermediary.

digetx committed:
libhost1x: handle tile'ness in host1x_gr2d_blit()

Pixel buffer format carries tiling info.

digetx committed:
libhost1x: align tiled pixel buffer pitch to 16 bytes

Fixes distorted display output with non-aligned framebuffer width.

digetx committed:
libhost1x: use pixel buffer depth info in host1x_gr2d_blit()

It should blit 16bit surfaces correctly now.

digetx committed:
libhost1x: make host1x_framebuffer/pixbuffer public

digetx committed:
libhost1x: make host1x_gr2d/3d_*() take pixbuffer instead of framebuffer

digetx committed:
libhost1x: make host1x_bo public

digetx committed:
libhost1x: use HW for detiling of a saved framebuffer

Use gr2d blit to detile pixel buffer.

digetx committed:
grate: drop grate_bo in favor of host1x_bo

digetx committed:
grate: replace get_*_framebuffer_bo with get_actual_framebuffer_pixbuf

digetx committed:
grate: refactor grate_3d_ctx_create_render_target

Replace it with grate_3d_ctx_bind_render_target(), dither state now belongs
to the target location, use pixel buffer.

digetx committed:
tests/grate: return double buffering

Grate API has all necessary bits for double buffering now.

digetx committed:
grate: refactor grate_3d_ctx_create_texture

Factor out texture state and use pixel buffer.

digetx committed:
libhost1x: blit if pitch/tiling mismatch in host1x_pixelbuffer_load_data()

And use that feature for texture data load in cube-textured test.

digetx committed:
tests/grate: cube-textured: print error message if texture load failed

To avoid confusion.

digetx committed:
tests: chdir to the grate dir in tests that use data/

So texture will be loaded in tests regardless of the current pwd.

digetx committed:
sync envytools-generated tgr_3d.xml.h

digetx committed:
grate: drop redundant FP counters reset in grate_3d_init()

The FP_UPLOAD_INST_ID_COMMON register should reset them at once.

kusma commented on src/libgrate/grate-2d.c:
```diff
@@ -47,8 +47,8 @@ void grate_clear(struct grate *grate)
 
-	err = host1x_gr2d_clear(gr2d, grate->fb->back, clear->r, clear->g,
-				clear->b, clear->a);
+	err = host1x_gr2d_clear(gr2d, grate->fb->back ?: grate->fb->front,
```
Huh, is `?:` a valid C operator? Some googling didn't make this clear...

digetx replied:
It is GCC extension, works fine in the UNIX world. https://gcc.gnu.org/onlinedocs/gcc-4.6.4/gcc/Conditionals.html

kusma commented on include/host1x.h:
```diff
@@ -66,6 +66,55 @@ struct host1x_gr3d;
+    PIX_BUF_FMT_D16_LINEAR_TILED    = PIX_BUF_FMT(7, 16, 1),
+    PIX_BUF_FMT_D16_NONLINEAR_TILED = PIX_BUF_FMT(8, 16, 1),
+    PIX_BUF_FMT_RGBA8888_TILED      = PIX_BUF_FMT(9, 32, 1),
+    PIX_BUF_FMT_RGBA_FP32_TILED     = PIX_BUF_FMT(10, 32, 1),
```
Hmm, I'm not loving mixing the tiling into the pixel format, as those are completely orthogonal features... Tiling is the order the pixels are stored, whereas the pixel format is the order the components of the pixels are stored. IIRC, the hardware even tiles on byte-sizes, as opposed to pixel-sizes.

I think we should instead have an `enum pixel_layout { PIX_BUF_LAYOUT_LINEAR = 0, PIX_BUF_LAYOUT_TILED = 1 }` or something.

digetx replied:
Factoring out layout format is also fine, I'll make it in the follow-up patch. We will probably end-up having multiple tiling formats because textures seems support several tiling formats and they differ from the display.

kusma replied:
Cool.

kusma commented on include/host1x.h:
```diff
@@ -66,6 +66,55 @@ struct host1x_gr3d;
+				struct host1x *host1x,
+				unsigned width, unsigned height,
+				unsigned pitch, enum pixel_format format);
+void host1x_pixelbuffer_free(struct host1x_pixelbuffer *pb);
```
I'm not 100% sure this is the best interface. We might want to allocate several miplevels in the same buffer, for instance. But, we don't need to do that *right now*, so maybe this is nice for the time being.

digetx replied:
I'm not sure too, certainly it will evolve :)

kusma commented on include/host1x.h:
```diff
@@ -66,6 +66,55 @@ struct host1x_gr3d;
+    PIX_BUF_FMT_D16_LINEAR    = PIX_BUF_FMT(7, 16, 0),
+    PIX_BUF_FMT_D16_NONLINEAR = PIX_BUF_FMT(8, 16, 0),
+    PIX_BUF_FMT_RGBA8888      = PIX_BUF_FMT(9, 32, 0),
+    PIX_BUF_FMT_RGBA_FP32     = PIX_BUF_FMT(10, 32, 0),
```
These should probably use `PIXEL_FORMAT_RGB565` etc from the rnndb-generated header, no?

digetx replied:
No, these formats are also used for the display framebuffer and they differ from the 3d stuff. The pixel buffer format is abstract, the corresponding hw value is picked by the pixel buffer user. The idea is that we can pass the "pixel buffer" around 3d <-> 2d <-> display without a hassle.

kusma replied:
Ah, I see.

kusma commented on src/libgrate/grate.c:
```diff
@@ -152,14 +152,10 @@ void grate_bind_framebuffer(struct grate *grate, struct grate_framebuffer *fb)
 }
 
-struct host1x_bo * grate_get_front_framebuffer_bo(struct grate_framebuffer *fb)
+struct host1x_pixelbuffer * grate_get_actual_framebuffer_pixbuf(
```
This is a small nit, but "actual" sounds a bit too abstract to me. How about using OpenGL terms (draw-buffer), and say `grate_get_draw_pixbuf` instead?

digetx replied:
No objections :)

kusma commented on src/libgrate/grate-3d.h:
```diff
@@ -0,0 +1,115 @@
+	struct grate_bo *bo;
+	unsigned pixel_format;
+	unsigned pitch;
+	bool dither_enable;
```
If we want to play by OpenGL-rules, dithering is not a render-target state, but a general render-state; switching render-target keeps the current dither-state.

digetx replied:
Yeah, I changed it in in the follow-up commit https://github.com/grate-driver/grate/pull/13/commits/c05db08346a2d9e194b6324a13be6883de57f1dd

kusma replied:
Ah, I see. If it was just a global state, we wouldn't be able to change it per target. So yeah, I think I agree.

kusma commented on src/libgrate/grate-3d-ctx.c:
```diff
@@ -423,70 +423,90 @@ void grate_3d_ctx_set_provoking_vtx_last(struct grate_3d_ctx *ctx, bool last)
-	tex->mip_filter = mip_filter;
-	tex->mag_filter = mag_filter;
-	tex->min_filter = min_filter;
+	ctx->textures[ctx->active_texture].max_lod = max_lod;
```
While this is the way OpenGL used to do things like this, bindless is now the new wine, where the texture object is specified directly. And especially for a low-level API, I think the following makes a lot more sense; perhaps we make it up to the callee to keep track of what object they want to modify (e.g pass `struct grate_texture *` to these functions instead of `struct grate_3d_ctx *`)? That would surely lead to less boiler-plate code...

digetx replied:
Sounds good to me, no objections.

kusma commented:
I forgot to say, very nice work! My nits are very minor, and I really like the result :)

digetx commented:
Thank you a lot for looking though the code! I'll add patches to this series addressing the comments.

kusma commented:
I really look forward to getting the cube-texture example transitioned to ASM-shaders, and we'll finally have a fully functional, blob-free, meaningful 3d example :)

digetx commented:
That should be easy now. For a really meaningful example we will need to add a depth buffer, because now the rendered cube is missing sides :)

kusma commented:
Yeah. Or just enable backface culling ;)

digetx commented:
I tried, doesn't help :/ But maybe I missed something.

kusma commented:
Hmm, seems culling doesn't really work. Either that, or we've messed up the vertex-order. I'll have a look.

kusma committed:
tests/grate: make triangle-winding consistent

This makes all triangles counter-clockwise, so we can start doing
backface culling.

kusma committed:
tests/grate: cull back-faces for cube-textured

digetx committed:
libhost1x: factor out pixel buffer layout

It's more consistent to have pixel format and data layout type separated.
Should be useful in a longer run as well, once we will have to add more
tiled layouts.

digetx committed:
grate: s/grate_get_actual_framebuffer_pixbuf/grate_get_draw_pixbuf/

To make it sound less abstract.

digetx committed:
grate: refactor texture handling a bit

Texture params (LOD level, MIP filter...) now belongs to the texture object
and not to the texture target. The texture creation/loading routines factored
out from cube-textured test to grate-texture.c.

digetx committed:
s/pb/pixbuf/ to avoid confusion with pushbuf

For better readability.

digetx committed:
libhost1x: be more verbose on error

In some cases we fail silently, in others completely ignoring the failure,
so at least printing a message in case of error should be handy.

digetx committed:
grate: handle single buffered vsync correctly

The display plane update ("show it" call in our case) causes indirect wait
for VBLANK, since the plane update blocks until VBLANK. Simply don't invoke
overlay/display show on grate_swap_buffers() in a single buffered mode
if grate vsync option is disabled (enabled by a "-v" cmd line argument).

digetx commented:
I added patches addressing the comments and couple more, please take a look. Also swapped texcoords in your "triangle-winding" commit, so cw sides are not inverted horizontally.

kusma reviewed (approved):
Looks good to me, thanks a lot!

digetx commented:
Nice!
